main/src/ca/uwaterloo/flix/language/phase/Typer.scala

```diff
@@ -1299,6 +1296,14 @@ object Typer {
           resultEff = declaredEff.getOrElse(actualEff)
         } yield (constrs, resultTyp, resultPur, resultEff)
 
+      case KindedAst.Expression.Upcast(exp, _, _, _, loc) =>
```

Comment:
@jaschdoc 

Here is a small mistake:

You should already have a tvar, pvar, and evar from the Upcast. These are the ones you want to unify with. You don't want to create fresh ones here. 

Then you want to do:

```
(constrs, tpe, pur, eff) <- visitExp(exp)
_ <- unifyM(tvar, tpe)
_ <- unifyM(pvar, pur)
_ <- unifyM(evar, eff)
````

Reply:
Yes but now the upcast expression gets the same type as the inner `exp`. The way I understand it is that we want the upcast expression to have the expected types of the environment (read: the expression enclosing the upcast expression) and the inner `exp` to just have its inferred type. In other words, upcast can be whatever is necessary.

In the test below `upcast f` should have type `Int32 -> Int32 & Impure` whereas `f` should have `Int32 -> Int32` because `if (true) (upcast f) else g` has type `Int32 -> Int32 & Impure`.

```flix
    def testUpcast03(): Unit =
      let f = x -> x + 1;
      let g = x -> { x + 1 as & Impure };
      let _ = if (true) (upcast f) else g;
      ()
```

Am I overlooking something, is there any way in the typer to unify with something else than the inner expression of upcast?

Reply:
Ah right- I was not thinking. In that case, drop the unification calls and just return the type variables (tvar, pvar, evar) etc.